File: appinventor/components/src/com/google/appinventor/components/runtime/ScatterChartDataModel.java
// -*- mode: java; c-basic-offset: 2; -*-
// Copyright 2019-2024 MIT, All rights reserved
// Released under the Apache License, Version 2.0
// http://www.apache.org/licenses/LICENSE-2.0

package com.google.appinventor.components.runtime;

import com.github.mikephil.charting.charts.ScatterChart;
import com.github.mikephil.charting.data.Entry;
import com.github.mikephil.charting.data.ScatterData;
import com.github.mikephil.charting.data.ScatterDataSet;
import com.github.mikephil.charting.interfaces.datasets.IScatterDataSet;
import com.github.mikephil.charting.utils.EntryXComparator;

import com.google.appinventor.components.common.PointStyle;

import com.google.appinventor.components.runtime.util.YailList;

import java.util.ArrayList;
import java.util.Collections;

/**
 * Handles the data operations & model-specific styling for Scatter
 * Chart data for the Chart component.
 * @see com.google.appinventor.components.runtime.ChartDataModel
 */
public class ScatterChartDataModel extends PointChartDataModel<
    Entry, IScatterDataSet, ScatterData, ScatterChart, ScatterChartView> {
  /**
   * Initializes a new ScatterChartDataModel object instance.
   *
   * @param data Chart data instance
   * @param view Scatter Chart View to link model to
   */
  public ScatterChartDataModel(ScatterData data, ScatterChartView view) {
    this(data, view, new ScatterDataSet(new ArrayList<Entry>(), ""));
  }

  protected ScatterChartDataModel(ScatterData data, ScatterChartView view,
      IScatterDataSet dataset) {
    super(data, view);
    this.dataset = dataset;
    this.data.addDataSet(dataset); // Safe add
    setDefaultStylingProperties();
  }

  @Override
  public void addEntryFromTuple(YailList tuple) {
    // Construct an entry from the prvoided tuple
    Entry entry = getEntryFromTuple(tuple);

    // If entry constructed successfully, add it to the Data Series
    if (entry != null) {
      /* TODO: The commented out line should be used, however, it breaks in certain cases.
         When this is fixed in MPAndroidChart, this method should use the commented method instead
         of the current implementation.
         See: https://github.com/PhilJay/MPAndroidChart/issues/4616
      */
      // getDataset().addEntryOrdered(entry);


      // In Line Chart based data series, the data is already pre-sorted.
      // We can thus run binary search by comparing with the x value, and
      // using an x+1 value to find the insertion point
      int index = Collections.binarySearch(entries, // Use the list of entries
          entry, // Search for the same x value as the entry to be added
          new EntryXComparator()); // Compare by x value

      // Value not found: insertion point can be derived from it
      if (index < 0) {
        // result is (-(insertion point) - 1)
        index = -index - 1;
      } else {
        // Get the entry count of the Data Set
        int entryCount = entries.size();

        // Iterate until an entry with a differing (higher) x value is found (this
        // is where the value should be inserted)
        // The reason for a loop is to pass through all the duplicate entries.
        while (index < entryCount && entries.get(index).getX() == entry.getX()) {
          index++;
        }
      }

      entries.add(index, entry);
    }
  }

  @Override
  protected void setDefaultStylingProperties() {
    if (dataset instanceof ScatterDataSet) {
      ((ScatterDataSet) dataset).setScatterShape(ScatterChart.ScatterShape.CIRCLE);
    }
  }

  /**
   * Changes the Point Shape of the Scatter Data Series.
   *
   * @param shape the desired point style
   */
  public void setPointShape(PointStyle shape) {
    if (!(dataset instanceof ScatterDataSet)) {
      return;
    }
    switch (shape) {
      case Circle:
        ((ScatterDataSet) dataset).setScatterShape(ScatterChart.ScatterShape.CIRCLE);
        break;

      case Square:
        ((ScatterDataSet) dataset).setScatterShape(ScatterChart.ScatterShape.SQUARE);
        break;

      case Triangle:
        ((ScatterDataSet) dataset).setScatterShape(ScatterChart.ScatterShape.TRIANGLE);
        break;

      case Cross:
        ((ScatterDataSet) dataset).setScatterShape(ScatterChart.ScatterShape.CROSS);
        break;

      case X:
        ((ScatterDataSet) dataset).setScatterShape(ScatterChart.ScatterShape.X);
        break;

      default:
        throw new IllegalArgumentException("Unknown shape type: " + shape);
    }
  }
}

File: appinventor/components-ios/src/ScatterChartDataModel.swift
// -*- mode: swift; swift-mode:basic-offset: 2; -*-
// Copyright © 2024 Massachusetts Institute of Technology, All rights reserved.
// Released under the Apache License, Version 2.0
// http://www.apache.org/licenses/LICENSE-2.0

import Foundation
import DGCharts

class ScatterChartDataModel: PointChartDataModel {
  var chartDataEntry: Array<ChartDataEntry> = []

  init(data: DGCharts.ScatterChartData, view: ScatterChartView) {
    super.init(data: data, view: view)
    let dataset = ScatterChartDataSet(entries: chartDataEntry, label: " ")
    self.dataset = dataset
    self.data.dataSets.append(dataset)
    setDefaultStylingProperties()
  }

  init(data: DGCharts.ScatterChartData, view: ScatterChartView, dataset: DGCharts.ChartDataSet) {
    super.init(data: data, view: view)
    self.dataset = dataset
    self.data.dataSets.append(dataset)
    setDefaultStylingProperties()
  }

  public override func addEntryFromTuple(_ tuple: YailList<AnyObject>) {
    // Assuming getEntryFromTuple is a function that returns an optional ChartDataEntry.
    // The tuple parameter type adjusted for Swift.
    guard let entry = getEntryFromTuple(tuple) else {
      // Not a valid entry
      return
    }
    // Assuming a correctly implemented binarySearch function that returns the index
    // where the entry should be inserted or the negative index - 1 if not found.
    var index = binarySearch(entry, chartEntries)
    if index < 0 {
      index = -(index + 1)
    } else {
      let entryCount = entries.count
      while index < entryCount && entries[index].x == entry.x  {
        index += 1
      }
    }

    // Insert the entry into the entries array.
    _entries.insert(entry, at: index)

    // Assuming you're updating some dataset that needs to be replaced entirely.
    // Performing UI updates asynchronously on the main thread.
    DispatchQueue.main.async {
      self.dataset?.replaceEntries(self.chartEntries)
    }
  }

  public override func setDefaultStylingProperties() {
    guard let scatterDataSet = dataset as? DGCharts.ScatterChartDataSet else {
      return
    }
    scatterDataSet.setScatterShape(.circle)
  }

  public func setPointShape(_ shape: PointStyle) {
    guard let scatterDataSet = dataset as? DGCharts.ScatterChartDataSet else {
      return
    }
    switch shape {
    case PointStyle.Circle:
      scatterDataSet.setScatterShape(.circle)
      break
    case PointStyle.Square:
      scatterDataSet.setScatterShape(.square)
      break
    case PointStyle.Triangle:
      scatterDataSet.setScatterShape(.triangle)
      break
    case PointStyle.Cross:
      scatterDataSet.setScatterShape(.cross)
      break
    case PointStyle.X:
      scatterDataSet.setScatterShape(.x)
      break
    default:
      break
    }
  }

}

